org/uacalc/alg/Malcev.java

```diff
@@ -732,6 +732,84 @@ public static Term fixedKEdgeTerm(SmallAlgebra alg, int k, ProgressReport report
     return null;
   }
 
+  /** 
+   * Alexandr Kazda's implementation (based on bits of previous UACalc code)
+   * of the local to global test for quasi weak near uninimity operation of
+   * arity <code>arity</code>. See the paper A. Kazda "Deciding the existence
+   * of quasi weak near unanimity terms in finite algebras"
+   *
+   * 
+   * @param testalg  an algebra 
+   * @param arity arity (at least 2)
+   * @return Boolean
+   */
+  public static boolean fixedKQwnu(SmallAlgebra testalg, int arity, ProgressReport report) {
+    if (arity < 2) throw new IllegalArgumentException("arity must be at least 2");
+    if (report != null) report.addStartLine("Using the local to global method to test for quasi WNU of arity " + arity + ". See the paper A. Kazda \"Deciding the existence of quasi weak near unanimity terms in finite algebras,\" 2020."); 
+    int power = arity;
+    BigProductAlgebra alg = new BigProductAlgebra(testalg, power);
+    final int size = testalg.cardinality();
+    //IntArray genvector = new IntArray(arity);
+    int a;
+    int b;
+    int i;
+    int j;
+    int[][] G= new int[arity][arity];
+    int[] F=new int[arity];
+    boolean found;
+    for (a=0;a<size;a++) {
+	    Arrays.fill(F,a);
+		    for(i=0;i<arity;i++)
+			    System.arraycopy(F,0,G[i],0,arity);//Fill the whole G with a's
+	    for(b=0;b<size;b++) {
+    		    List<IntArray> gens = new ArrayList<IntArray>(arity);
+		    if (a==b)
+			    break;
+		    for(i=0;i<arity;i++) {
+			    	G[i][i]=b; //The diagonal of G is b
+			    	gens.add(new IntArray(G[i]));
+		    }
+		    Closer closer = new Closer(alg, gens);
+        	    closer.setSuppressOutput(true);
+        	    List<IntArray> sub = closer.sgClose();
+		    found=false;
+		 	//Look for element of the form (q,q,...,q) in sub
+		    for (i=0;i<size;i++) {
+ 				Arrays.fill(F,i);
+				IntArray Aarray = new IntArray(F);
+				if (sub.contains(Aarray)) {
+					found=true;
+					break;
+				}
+		    }
+        	    if (!found) {
+        		  if (report != null) {
+        		    report.addLine("No diagonal element in the subuniverse generated by the tuples:");
+        		    for (int k = 0; k < arity; k++) {
+        		      String GenString = Arrays.toString(G[k]);
+        		      report.addLine(GenString);
+        		    }
+        		    report.addEndingLine("There is no " + arity + "-ary quasi WNU.");
+        		  }
+        		  else {
+        		    report.addLine("No diagonal element in the subuniverse generated by the tuples:");
+        		    for (int k = 0; k < arity; k++) {
+        		      String GenString = Arrays.toString(G[k]);
+        		      System.out.println(GenString);
+        		    }
+        		    System.out.println("The subuniverse is " + sub.toString());
+        		  }
+        		  return false;
```

Comment:
This will short-circuit the function if `report` is `null`.

Reply:
This is supposed to short-circuit the function whenever `found==false`, regardless of `report`. I think that the line is doing just that.

Reply:
Got it.

Reply:
Hi Alexandr and Matthew, 
Thanks for the good work. After another day I will do a merge, partly so you 
can make changes and partly since I always have to read up on git each
time before I use it.  --ralph